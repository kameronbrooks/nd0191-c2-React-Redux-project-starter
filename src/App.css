body {
  background-color: #f8f9fa;

}

.login-form {
  background-color: rgb(214, 221, 230);
  padding: 10px 10px 10px 10px;
  margin-top: 30px;
  border-radius: 10px;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);
  border: 1px solid #d3d3d3;

  color: #222;
}

.hint-text {
  color: #777;
  font-style: italic;
  text-align: center;
  padding-bottom: 15px;
  margin-bottom: 15px;
  border-bottom: 1px solid #e6e6e6;
}


.navbar {
  background-color: #222;
  padding: 0px 0px 0px 0px;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);
  border: 1px solid #d3d3d3;
  color: #ffffff;
  display: flex;
  justify-content: space-between;
  position: relative;
  height: 50px;
}

.avatar-box {
  position: absolute;
  top: 0px;
  right: 10px;
  transform: translateY(10px);
  background-color: #ffffff;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);

  border: 1px solid #d3d3d3;
  color: #222;
  float: right;
  width: 65px;
  height: 65px;
  border-radius: 7px 7px 7px 7px;

  background-size: cover;
  transition: transform 0.3s, box-shadow 0.3s;
}

.avatar-box:hover {

  transform: scale(1.07) translateY(7px);
  box-shadow: 0 8px 10px -4px rgba(0, 0, 0, 0.2);
}

.avatar-box .logout-link {
  position: absolute;
  bottom: 0px;
  right: 0px;
  left: 0px;
  text-align: center;
  color: #fff;
  background-color: rgba(0, 0, 0, 0.6);
  font-size: 75%;
  border-top: 1px solid rgba(255, 255, 255, 0.5);
  border-radius: 0 0 7px 7px;
  cursor: pointer;
  opacity: 0;
  transition: opacity 0.3s;
}

.avatar-box:hover .logout-link {
  opacity: 1;
}

.navlinks-div {
  margin-left: 20px;
  padding: 3px 40px 3px 3px;
  display: flex;
  justify-content: space-between;
  width: 30%;
  color: #fff;
  border-right: rgba(255, 255, 255, 0.3) 1px solid;
  height: 42.5px;
}

.navlinks-div a {
  color: #ffffff;
  box-sizing: border-box;
}

.navlinks-div .active {
  box-sizing: border-box;
  border-bottom: 4px solid #ffffff;
}

.bucket-head {
  border-bottom: 1px solid #444;
  text-align: center;
  background-color: rgb(34, 34, 34);
  border-radius: 5px 5px 0 0;
  color: #fff;
  font-weight: 300;
  font-size: 75%;
}

.bucket-head h2 {
  font-weight: 300;
}

.bucket-body {
  display: flex;
  flex-wrap: wrap;
  justify-content: start;

}

.bucket-col {
  flex: 0 0 32%;
  max-width: 32%;
  margin: 0.6665%
}

.question-bucket {
  border: 1px solid #444;
  margin: 7px 7px 7px 7px;
  border-radius: 5px 5px 5px 5px;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);

}

.spinner-container {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  /* Full height of the parent */
  width: 100%;
  /* Full width of the parent */
  min-height: 100px;
}

.question {
  padding: 10px 10px 10px 10px;
  border: 1px solid #444;
  border-radius: 5px 5px 5px 5px;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);
}

.justify-center {
  display: flex;
  justify-content: center;
}

.detail-text {
  color: #777;
  font-style: italic;
  text-align: center;
  padding-bottom: 15px;
  margin-bottom: 15px;
  border-bottom: 1px solid #aaa;
}

.question-nav {
  display: flex;
  justify-content: space-between;
  padding: 10px 10px 10px 10px;

}

.author-img {
  width: 200px;
  height: 200px;
  border-radius: 50%;
  margin-right: 10px;
}

.answer-button {
  width: 100%;
}

.results-div {
  padding: 10px 10px 10px 10px;
  border: #444 1px solid;
  border-radius: 5px 5px 5px 5px;
  box-shadow: 0 10px 7px -4px rgba(0, 0, 0, 0.2);
}

.results-div > .title {
  text-align: center;
  border-bottom: #444 1px solid;
  border-radius: 5px 5px 0 0;
}

.results-body {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.results-details {
  flex: 0 0 50%;
  max-width: 50%;
  padding: 10px 10px 10px 10px;
}
.results-chart{
  flex: 0 0 50%;
  max-width: 50%;
  padding: 10px 10px 10px 10px;
  display: flex;
  justify-content: center;
}
