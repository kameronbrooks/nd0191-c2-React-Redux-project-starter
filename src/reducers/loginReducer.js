
import { LOGIN_USER, LOGOUT_USER, REQUEST_USER_LOGIN, LOGIN_USER_ERROR} from "../actions/loginActions";

const initialState = {
  user: null,
  isLoggedIn: false,
  loginError: false,
  loginErrorMessage: null,
  loading: false
};

const loginReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_USER_LOGIN:        // Dispatched when the user clicks the login button
      return {
        ...state,
        loginError: false,
        loginErrorMessage: null,
        loading: true
      };
    case LOGIN_USER:                // Dispatched when the user is successfully logged in
      return {
        ...state,
        user: action.user,
        isLoggedIn: true,
        loginError: false,
        loginErrorMessage: null,
        loading: false
      };
    case LOGOUT_USER:               // Dispatched when the user clicks the logout button
      return {
        ...state,
        user: null,
        isLoggedIn: false
      };
    case LOGIN_USER_ERROR:          // Dispatched when the user enters an invalid username or password
      return {
        ...state,
        loginError: true,
        loginErrorMessage: "Invalid username or password",
        loading: false
      };
    default:
      return state;
  }
};

export default loginReducer;