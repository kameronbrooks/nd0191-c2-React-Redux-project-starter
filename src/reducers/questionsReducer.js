import {
  REQUEST_QUESTIONS,
  RECEIVE_QUESTIONS,
  REQUEST_ADD_QUESTION,
  REQUEST_QUESTIONS_ERROR,
  ADD_QUESTION,
  ADD_QUESTION_ERROR,
  ANSWER_QUESTION,
  ANSWER_QUESTION_ERROR,
  DELETE_QUESTION
} from "../actions/questionActions";


const initialState = {
  questions: {},
  loading: false,
  error: false,
  errorMessage: null,
  savingNewQuestion: false,
};

const questionsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_QUESTIONS:
      return {
        ...state,
        loading: true,
        error: false,
        errorMessage: null
      };
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        questions: action.questions,
        loading: false,
        error: false,
        errorMessage: null
      };
    case REQUEST_QUESTIONS_ERROR:
      return {
        ...state,
        loading: false,
        error: true,
        errorMessage: "Error loading questions"
      };
    case REQUEST_ADD_QUESTION:
      return {
        ...state,
        savingNewQuestion: true,
        error: false,
        errorMessage: null
      };
    case ADD_QUESTION:
      return {
        ...state,
        questions: {...state.questions,
          [action.question.id]: action.question
        },
        savingNewQuestion: false
      };
    case ADD_QUESTION_ERROR:
      return {
        ...state,
        error: true,
        errorMessage: "Error adding question",
        savingNewQuestion: false
      };
    case ANSWER_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.quid]: {
            ...state.questions[action.quid],
            [action.answer]: {
              ...state.questions[action.quid][action.answer],
              votes: state.questions[action.quid][action.answer].votes.concat([action.authedUser])
            }
          }
        }
      };
    case ANSWER_QUESTION_ERROR:
      return {
        ...state,
        error: true,
        errorMessage: "Error answering question"
      };
    case DELETE_QUESTION:
      return {
        ...state,
        questions: state.questions.filter(question => question.id !== action.question.id)
      };
    default:
      return state;
  }
}

export default questionsReducer;