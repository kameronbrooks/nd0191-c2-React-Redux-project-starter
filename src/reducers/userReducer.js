import { REQUEST_USERS, RECEIVE_USERS, REQUEST_USERS_ERROR } from "../actions/userActions";

const initialState = {
  users: {},
  loading: false,
  error: false,
  errorMessage: null
};

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_USERS:
      return {
        ...state,
        loading: true,
        error: false,
        errorMessage: null
      };
    case RECEIVE_USERS:
      return {
        ...state,
        users: action.users,
        loading: false,
        error: false,
        errorMessage: null
      };
    case REQUEST_USERS_ERROR:
      return {
        ...state,
        loading: false,
        error: true,
        errorMessage: "Error loading users"
      };
    default:
      return state;
  }
};

export default userReducer;