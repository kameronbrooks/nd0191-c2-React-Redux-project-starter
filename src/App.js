import logo from './logo.svg';
import './App.css';
import {Context} from './Context';
import {BrowserRouter as Router, Routes, Route} from 'react-router-dom';
import LoginPage from './components/LoginPage';
import HomePage from './components/HomePage';
import LeaderboardPage from './components/LeaderboardPage';
import QuestionPage from './components/QuestionPage';


const App = () => {
  console.log('App');
  return (
    <div className="App">
      <Router>
        <Routes>
          <Route path="/login" element={<LoginPage />} />
          <Route path="/" element={<HomePage />} />
          <Route path="/leaderboard" element={<LeaderboardPage />} />
          <Route path="/questions/:id" element={<QuestionPage />} />
          <Route path="/questions/new" element={<QuestionPage />} />
        </Routes>
      </Router>
    </div>
  );
}

export default App;
