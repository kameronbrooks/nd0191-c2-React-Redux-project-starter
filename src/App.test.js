import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import App from './App';
import { combineReducers, createStore } from 'redux';

import 'bootstrap/dist/css/bootstrap.min.css';
import loginReducer from './reducers/loginReducer';
import questionsReducer from './reducers/questionsReducer';
import userReducer from './reducers/userReducer';


// This function will render a component with a fresh store
const renderWithFreshStore = (ui) => {
  const store = createStore(
    combineReducers({ 
      login: loginReducer,
      questions: questionsReducer,
      users: userReducer
    })
  );
  return render(
    <Provider store={store}>
      {ui}
    </Provider>
  );
}

// ==================================
//        App Component Tests
// ==================================
describe('App', () => {
  it('redirects to the login page', () => {
    renderWithFreshStore(
        <App />
    );
  
    const h1Element = screen.getByText(/Login Page/i);
    expect(h1Element).toBeInTheDocument();
  });
});

describe('App', () => {
  it('redirects to the home page upon successful login', async () => {
    const {container} = renderWithFreshStore(
      <App />
    );
  
    const usernameInput = container.querySelector('#username-input');
    const passwordInput = container.querySelector('#password-input');

    fireEvent.change(usernameInput, { target: { value: 'zoshikanlu' } });
    fireEvent.change(passwordInput, { target: { value: 'pass246' } });

    const loginButton = container.querySelector('#login-button');

    fireEvent.submit(loginButton);

    await waitFor(() => expect(screen.getByText(/New Questions/i)).toBeInTheDocument(), {timeout: 30000});

  });
});

describe('App', () => {
  it('shows error on unsuccessful login', async () => {
    const {container} = renderWithFreshStore(
      <App />
    );
  
    const usernameInput = container.querySelector('#username-input');
    const passwordInput = container.querySelector('#password-input');

    fireEvent.change(usernameInput, { target: { value: 'zoshikanla' } });
    fireEvent.change(passwordInput, { target: { value: 'pass247' } });

    const loginButton = container.querySelector('#login-button');

    fireEvent.submit(loginButton);

    await waitFor(() => expect(screen.getByText(/Invalid username or password/i)).toBeInTheDocument(), {timeout: 30000});

  });
});

describe('App', () => {
  it('will match snapshot', () => {
    const {container} = renderWithFreshStore(
      <App />
    );
    expect(container).toMatchSnapshot();
  });
});
