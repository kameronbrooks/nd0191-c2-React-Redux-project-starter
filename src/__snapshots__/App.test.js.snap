// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App will match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div>
      <h1>
        Login Page
      </h1>
      <div
        class="container"
      >
        <form
          class="login-form"
          novalidate=""
        >
          <h3>
            Login
          </h3>
          <hr />
          <div
            class="mb-3 input-group has-validation"
          >
            <span
              class="input-group-text"
            >
              Login
            </span>
            <input
              class="form-control"
              id="username-input"
              placeholder="Username"
              required=""
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Please choose a username.
            </div>
          </div>
          <div
            class="mb-3 input-group has-validation"
          >
            <span
              class="input-group-text"
            >
              Password
            </span>
            <input
              class="form-control"
              id="password-input"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Please choose a password.
            </div>
          </div>
          <hr />
          <button
            class="ms-auto btn btn-dark"
            id="login-button"
            type="submit"
          >
            Login
          </button>
        </form>
        <br />
        <div
          class="hint-text"
        >
          Use 
          <strong>
            tylermcginnis
          </strong>
           as the username and 
          <strong>
            abc321
          </strong>
           for the password
        </div>
      </div>
    </div>
  </div>
</div>
`;
